### ISSUE TITLE
hello from lib03d3xx - catkin_make having cv issues

### ISSUE THREAD

**Issue by wegunterjr** [lovepark/o3d3xx-ros#8]
```
viki@c3po:~/dev/o3d3xx-catkin$ catkin_make verbose=1
Base path: /home/viki/dev/o3d3xx-catkin
Source space: /home/viki/dev/o3d3xx-catkin/src
Build space: /home/viki/dev/o3d3xx-catkin/build
Devel space: /home/viki/dev/o3d3xx-catkin/devel
Install space: /home/viki/dev/o3d3xx-catkin/install
####
#### Running command: "make cmake_check_build_system" in "/home/viki/dev/o3d3xx-catkin/build"
####
####
#### Running command: "make verbose=1 -j2 -l2" in "/home/viki/dev/o3d3xx-catkin/build"
####
[  0%] Built target std_msgs_generate_messages_cpp
[  0%] Built target _o3d3xx_generate_messages_check_deps_Rm
[  0%] Built target _o3d3xx_generate_messages_check_deps_Dump
[  0%] Built target _o3d3xx_generate_messages_check_deps_Config
[  0%] Built target _o3d3xx_generate_messages_check_deps_GetVersion
[  0%] Built target std_msgs_generate_messages_lisp
Linking CXX executable /home/viki/dev/o3d3xx-catkin/devel/lib/o3d3xx/o3d3xx_file_writer_node
[  0%] Built target std_msgs_generate_messages_py
[ 25%] Built target o3d3xx_generate_messages_cpp
[ 50%] Built target o3d3xx_generate_messages_lisp
[ 81%] Built target o3d3xx_generate_messages_py
[ 87%] Built target o3d3xx_node
Scanning dependencies of target o3d3xx_config_node
[ 93%] CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `cv::operator<<(cv::FileStorage&, charBuilding CXX object o3d3xx/CMakeFiles/o3d3xx_config_node.dir/src/o3d3xx_config_node.cpp.o
 const*)':
o3d3xx_file_writer_node.cpp:(.text+0xa5): undefined reference to `cv::operator<<(cv::FileStorage&, cv::String const&)'
CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `cv::FileStorage& cv::operator<< <cv::Mat>(cv::FileStorage&, cv::Mat const&)':
o3d3xx_file_writer_node.cpp:(.text+0x282): undefined reference to `cv::error(int, cv::String const&, char const*, char const*, int)'
o3d3xx_file_writer_node.cpp:(.text+0x2a9): undefined reference to `cv::write(cv::FileStorage&, cv::String const&, cv::Mat const&)'
CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `cv::String::String(char const*)':
o3d3xx_file_writer_node.cpp:(.text._ZN2cv6StringC2EPKc[_ZN2cv6StringC5EPKc]+0x4f): undefined reference to `cv::String::allocate(unsigned long)'
CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `cv::String::~String()':
o3d3xx_file_writer_node.cpp:(.text._ZN2cv6StringD2Ev[_ZN2cv6StringD5Ev]+0x14): undefined reference to `cv::String::deallocate()'
CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `cv::String::String(std::string const&)':
o3d3xx_file_writer_node.cpp:(.text._ZN2cv6StringC2ERKSs[_ZN2cv6StringC5ERKSs]+0x69): undefined reference to `cv::String::allocate(unsigned long)'
CMakeFiles/o3d3xx_file_writer_node.dir/src/o3d3xx_file_writer_node.cpp.o: In function `O3D3xxFileWriterNode::ImageCb(boost::shared_ptr<sensor_msgs::Image_<std::allocator<void> > const> const&, std::string const&)':
o3d3xx_file_writer_node.cpp:(.text._ZN20O3D3xxFileWriterNode7ImageCbERKN5boost10shared_ptrIKN11sensor_msgs6Image_ISaIvEEEEERKSs[_ZN20O3D3xxFileWriterNode7ImageCbERKN5boost10shared_ptrIKN11sensor_msgs6Image_ISaIvEEEEERKSs]+0x4d8): undefined reference to `cv::FileStorage::FileStorage(cv::String const&, int, cv::String const&)'
o3d3xx_file_writer_node.cpp:(.text._ZN20O3D3xxFileWriterNode7ImageCbERKN5boost10shared_ptrIKN11sensor_msgs6Image_ISaIvEEEEERKSs[_ZN20O3D3xxFileWriterNode7ImageCbERKN5boost10shared_ptrIKN11sensor_msgs6Image_ISaIvEEEEERKSs]+0x5fc): undefined reference to `cv::imwrite(cv::String const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [/home/viki/dev/o3d3xx-catkin/devel/lib/o3d3xx/o3d3xx_file_writer_node] Error 1
make[1]: *** [o3d3xx/CMakeFiles/o3d3xx_file_writer_node.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Linking CXX executable /home/viki/dev/o3d3xx-catkin/devel/lib/o3d3xx/o3d3xx_config_node
[ 93%] Built target o3d3xx_config_node
make: *** [all] Error 2
Invoking "make verbose=1 -j2 -l2" failed
viki@c3po:~/dev/o3d3xx-catkin$ 
```

**Comment by tpanzarella** [comment]
This is again a linker error related to your OpenCV install (see: lovepark/libo3d3xx#18).

I recommend that you remove the manual OpenCV install that you did and just use the binary install from apt. In fact, you'll get the correct version by doing your ROS install: `sudo apt-get install ros-indigo-desktop-full` as OpenCV is a dependency. IIRC, ROS Indigo uses OpenCV 2.4 by default -- i.e., all of the OpenCV related packages (e.g., `cv_bridge`) are linked against 2.4.

Also note that `o3d3xx-ros` has only been tested with ROS Indigo (hence my recommendation for `indigo` in the install command above). Are you using Indigo? Jade? Other? I am recommending Ubuntu 14.04 with Indigo as they are both LTS releases.

Bottom line: Unless your algorithm or application really needs something that OpenCV 3 is uniquely providing, I think you should seriously consider reevaluating your base setup to use the supported binaries, tested packages, and LTS releases. Particularly if you are building production code, which, most people using the O3D3xx sensor are doing.

I'm happy to help further if I can, but I do not have your setup (i.e., virtual machine, source build of OpenCV 3.0 running in parallel with 2.4 binaries, etc.) to test against.

**Comment by wegunterjr** [comment]
thanks for the suggestion. i went ahead and just installed ros-indigo-desktop-full in a new virtual machine *(Ubuntu 14.04) and then proceeded to install the driver for lib03d3xx and then the ROS install. I am currently stuck on this:
wegunterjr@ros-VB:~/dev$ roslaunch o3d3xx camera.launch
[camera.launch] is neither a launch file in package [o3d3xx] nor is [o3d3xx] a launch file name
The traceback for the exception was written to the log file

**Comment by tpanzarella** [comment]
Be sure you followed the instructions [here](https://github.com/lovepark/o3d3xx-ros#building-and-installing-the-software). Specifically this (adjust for your installation paths):

```
if [ -f /opt/ros/indigo/setup.bash ]; then
    source /opt/ros/indigo/setup.bash
fi

cd ${HOME}

export LPR_ROS=${HOME}/ros

if [ -d ${LPR_ROS} ]; then
    for i in $(ls ${LPR_ROS}); do
        if [ -d ${LPR_ROS}/${i} ]; then
            if [ -f ${LPR_ROS}/${i}/setup.bash ]; then
                source ${LPR_ROS}/${i}/setup.bash --extend
            fi
        fi
    done
fi
```

**Comment by tpanzarella** [comment]
In follow up to above, I'd also point out that this should go into your `~/.bashrc` or `~/.bash_profile` as appropriate. If trying to run the launch file just after executing the install, you likely would need to manually source in the `~/.bashrc` or `~/.bash_profile` if you are not opening a new shell to launch the camera driver. If using `~/.bash_profile` (as I do) make sure your terminal runs as a "login shell".

**Comment by tpanzarella** [comment]
If you have everything installed correctly, you should be able to do something simple like:

```
$ roscd o3d3xx
```

And it will take you into the `share` folder of the installation. If you cannot get that far, the issue is the environment setup, nothing specific to the camera.

**Comment by wegunterjr** [comment]
hmmm..okay, but with my ubuntu install i have a ~/.profile. I don't have a bash_profile.  Should i add it, and put all the references in it?

**Comment by tpanzarella** [comment]
`~/.profile`

**Comment by wegunterjr** [comment]
ok.  Here is my ~/.profile contencts.

# ~/.profile: executed by the command interpreter for login shells.
# This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login
# exists.
# see /usr/share/doc/bash/examples/startup-files for examples.
# the files are located in the bash-doc package.

# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.
#umask 022

# if running bash
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
        . "$HOME/.bashrc"
    fi
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin:$PATH"
fi

if [ -f /opt/libo3d3xx/etc/setup.bash ]; then
    source /opt/libo3d3xx/etc/setup.bash
fi

if [ -f /opt/ros/indigo/setup.bash ]; then
    source /opt/ros/indigo/setup.bash
fi

cd ${HOME}

export LPR_ROS=${HOME}/ros

if [ -d ${LPR_ROS} ]; then
    for i in $(ls ${LPR_ROS}); do
        if [ -d ${LPR_ROS}/${i} ]; then
            if [ -f ${LPR_ROS}/${i}/setup.bash ]; then
                source ${LPR_ROS}/${i}/setup.bash --extend
            fi
        fi
    done
fi

**Comment by tpanzarella** [comment]
OK. So, can you `roscd o3d3xx` from anywhere w/o issue?

**Comment by wegunterjr** [comment]
negative...I thought I could, but can't right now.

**Comment by tpanzarella** [comment]
Can you verify that the code actually got installed into `~/ros` as your `~/.profile` suggests that it should. If not, please be sure that you specified the `CMAKE_INSTALL_PREFIX` at install time:

```
$ cd ~/dev/o3d3xx-catkin
$ catkin_make
$ catkin_make -DCMAKE_INSTALL_PREFIX=${LPR_ROS}/o3d3xx install
```

**Comment by wegunterjr** [comment]
contents of my ~/ros folder:

'''wegunterjr@ros-VB:~/ros/o3d3xx$ ls -asl
total 52
 4 drwxrwxr-x 5 wegunterjr wegunterjr  4096 Oct 26 16:18 .
 4 drwxrwxr-x 3 wegunterjr wegunterjr  4096 Oct 26 15:43 ..
 0 -rw-rw-r-- 1 wegunterjr wegunterjr     0 Oct 26 15:43 .catkin
 4 -rwxr-xr-x 1 wegunterjr wegunterjr   506 Oct 26 15:38 env.sh
 4 drwxrwxr-x 3 wegunterjr wegunterjr  4096 Oct 26 15:52 include
 4 drwxrwxr-x 5 wegunterjr wegunterjr  4096 Oct 26 15:52 lib
 4 -rw-r--r-- 1 wegunterjr wegunterjr    67 Oct 26 16:18 .rosinstall
 4 -rw-r--r-- 1 wegunterjr wegunterjr   260 Oct 26 15:38 setup.bash
 4 -rw-r--r-- 1 wegunterjr wegunterjr  2540 Oct 26 16:18 setup.sh
12 -rwxr-xr-x 1 wegunterjr wegunterjr 12271 Oct 26 16:18 _setup_util.py
 4 -rw-r--r-- 1 wegunterjr wegunterjr   270 Oct 26 15:38 setup.zsh
 4 drwxrwxr-x 4 wegunterjr wegunterjr  4096 Oct 26 15:52 share
wegunterjr@ros-VB:~/ros/o3d3xx$ 
'''

**Comment by tpanzarella** [comment]
In your current shell, execute:

```
. ~/ros/o3d3xx/setup.bash --extend
```

Once you do that, see if you can `roscd o3d3xx`. If you can, see if you can `roslaunch o3d3xx camera.launch`.

**Comment by wegunterjr** [comment]
Okay...while in my ~/ros folder, i executed what you listed above, and I was able to roscd and roslaunch.  Do i need to add something somewhere so I can get to it all the time?

**Comment by tpanzarella** [comment]
Well, that is what the code in your `~/.profile` is supposed to do. So, one of two things:

1. There is a typo in that code
2. That code is not being run

Per 2 above, make sure your terminal program runs as a "login" shell. For example in the `gnome-terminal`:

![screenshot from 2015-10-26 16 50 57](https://cloud.githubusercontent.com/assets/645771/10742296/d071aac6-7c02-11e5-980f-487a013bf21d.png)

**Comment by wegunterjr** [comment]
aha!  that seems to be it. I closed all terminals and opened a new one after making the change and I can:
'''roscd o3d3xx'''

**Comment by tpanzarella** [comment]
OK. Good. Sounds like you are up and running. Let's close this for now and if anything else comes up, we should open that as a new issue.
